src/worker/mod.rs

```diff
@@ -369,51 +412,86 @@ impl<
+                //  - we got an error from heartbeat()
+                self.terminated = exit.is_ok();
+
+                if let Ok(true) = exit {
+                    let running = self.force_fail_all_workers("terminated").await;
+                    if running != 0 {
+                        return Ok((RunCeaseReason::FaktoryInstruction, running));
+                    }
+                }
 
-        // there are a couple of cases here:
-        //
-        //  - we got TERMINATE, so we should just return, even if a worker is still running
-        //  - we got TERMINATE and all workers have exited
-        //  - we got an error from heartbeat()
-        //
-        self.terminated = exit.is_ok();
+                // we want to expose worker errors, or otherwise the heartbeat error
+                let mut results = Vec::with_capacity(nworkers);
+                while let Some(res) = join_set.join_next().await {
+                    results.push(res.expect("joined ok"));
+                }
+                let results = results.into_iter().collect::<Result<Vec<_>, _>>();
 
-        if let Ok(true) = exit {
-            let running = self.force_fail_all_workers().await;
-            if running != 0 {
-                return Ok(running);
+                match exit {
+                    Ok(_) => results.map(|_| (RunCeaseReason::FaktoryInstruction, 0)),
+                    Err(e) => results.and(Err(e)),
+                }
             }
-        }
-
-        // we want to expose worker errors, or otherwise the heartbeat error
-        let mut results = Vec::with_capacity(workers_count);
-        while let Some(res) = join_set.join_next().await {
-            results.push(res.expect("joined ok"));
-        }
-
-        let result = results.into_iter().collect::<Result<Vec<_>, _>>();
-
-        match exit {
-            Ok(_) => result.map(|_| 0),
-            Err(e) => result.and(Err(e)),
-        }
+        };
+        report
     }
 
-    /// Run this worker until the server tells us to exit or a connection cannot be re-established.
+    /// Run this worker until the server tells us to exit or a connection cannot be re-established,
+    /// or a signal from the user-space code has been received via a future passed to [`WorkerBuilder::with_graceful_shutdown`]
     ///
-    /// This function never returns. When the worker decides to exit, the process is terminated.
+    /// This function never returns. When the worker decides to exit or a signal to shutdown gracefully
+    /// has been received, the process is terminated within the [shutdown period](WorkerBuilder::shutdown_timeout).
     pub async fn run_to_completion<Q>(mut self, queues: &[Q]) -> !
     where
         Q: AsRef<str>,
     {
-        use std::process;
-        while self.run(queues).await.is_err() {
-            if self.reconnect().await.is_err() {
-                break;
-            }
-        }
+        let sig = self.shutdown_signal.take();
+        tokio::select! {
```

Comment:
Why can't this just call `self.run` and allow `run` to deal with the graceful shutdown logic? Especially if we make the change around `.take()` above that I suggested.

Reply:
We really could not with `take()`, but it's now possible.